### DISCUSSION TITLE
Warm up period

### DISCUSSION THREAD

**Discussion by tubular2000** [kernc/backtesting.py#321] [General]
hi, is it possible to add a warm up period when calling run()? if my indicator needs 500 bars to start generating an output, i don't want these bars to be included in the backtest range.

**Comment by kernc** [comment]
We do skip the first couple bars, precisely to avoid testing on NaN indicators:
https://github.com/kernc/backtesting.py/blob/0b2325f94ff5dbe3486202a08cd981a3cd123d41/backtesting/backtesting.py#L1144-L1147
By _backtest range_, do you mean `stats['End']` and `stats['Duration']`?

**Reply by tubular2000** [reply]
what i mean would be to be able to call run like that:
bt.run(start=2020-01-01, end=2021-01-01) or bt.run(start=2020-01-01, numbar=500).
so if i have 100 000 bar in my dataset i can easily run backtest on subset of data without redefining the dataset every time. if i start the backtest at bar 1000, my indicator would still see these previous 1000 bars, but they would not be included in the stats.

**Reply by kernc** [reply]
You could simply pass a shortened data frame to the backtest instance?
```py
bt = Backtest(df.iloc[1000:10000], ...)  # Backtest bars 1000..10000
```

**Reply by tubular2000** [reply]
i could, but my indicators would not see df[0:1000], and if i include this range in the backtest it will influence the stats

**Reply by kernc** [reply]
Even if you include this range, it _won't_ influence the stats if your indicators _aren't forward-looking_ (as they shouldn't be). For instance, with `SMA(df.Close, 20)`, the first 19 result values will be equal to `NaN` and thus skipped by the code [snippet above](https://github.com/kernc/backtesting.py/discussions/321#discussioncomment-667119).

**Reply by tubular2000** [reply]
thank you for your help. i managed to make it work by precomputing my indicators in the main dataframe and backtesting a subset like you proposed.

thanks for making this library available to us!

**Reply by mikeFyngers** [reply]
Thanks kernc and all for the great package!

I implemented what I understand is the suggested solution but I have a major drawback now. I cannot use the class parameters anymore for optimization. Can I circumvent this?

I implemented a long term version of the SMA cross over in the first tutorial, see https://kernc.github.io/backtesting.py/doc/examples/Quick%20Start%20User%20Guide.html. DataFrame `prices` has the OHLC data from which I precompute the indicator data. I then prune the `prices` to remove the warm-up period because in reality I will have the indicator available already on day one. I define a `PassThrough` function to ingest the precomputed data via the suggested indicator definition via `self.I()`. However, the class variables `n1` and `n2` are then not used at all. Is there an alternative to including a second `DataFrame`on top of `data.self` with additional 200 days prepended (in this case). I understand that this opens the door for nasty issues of misaligning data but my current solution isnt satisfactory either. Thanks!

```
import pandas as pd

def SMA(values, n):
    """
    Return simple moving average of `values`, at
    each step taking into account `n` previous values.
    """
    return pd.Series(values).rolling(n).mean()

def PassThrough(values):
    """
    Return unchanged `values`, required for call to Indicator creation on Strategy.
    """
    return values

prices=prices.assign(sma200c=SMA(prices.Close, 200))
prices=prices.assign(sma50c=SMA(prices.Close, 50))

# Remove the first 200 rows such that we have a data set without warm-up
prices=prices.iloc[slice(199, len(prices))]

from backtesting import Strategy
from backtesting.lib import crossover

class SmaCross(Strategy):
    # Define the two MA lags as *class variables*
    # for later optimization
    n1 = 50 # Fast
    n2 = 200 # Slow
    
    def init(self):
        # Get precomputed data
        self.sma1 = self.I(PassThrough, self.data.sma50c, name='SMA 50c', overlay=True)
        self.sma2 = self.I(PassThrough, self.data.sma200c, name='SMA 200c', overlay=True)
    
    def next(self):
        #print(self.sma1)
        # If sma1 (fast) crosses above sma2 (slow), close any existing
        # short trades, and buy the asset
        if crossover(self.sma1, self.sma2):
            self.position.close()
            self.buy()

        # Else, if sma1 (fast) crosses below sma2 (slow) , close any existing
        # long trades, but dont sell the asset
        # See also https://github.com/kernc/backtesting.py/discussions/353        
        elif crossover(self.sma2, self.sma1):
            self.position.close()
            
        elif not self.position and self.sma1[-1]>self.sma2[-1]:
            # When strategy starts and there is no position yet, buy 
            # if today's sma1 > sma2 (instead of waiting for first crossing)
            self.buy()

from backtesting import Backtest

bt = Backtest(prices, SmaCross, cash=10_000, commission=.002)
stats = bt.run()
stats['_trades']
```

**Reply by kernc** [reply]
@mikeFyngers If you simply did:
```py
        self.sma1 = self.I(SMA, self.data.Close, 50, overlay=True)
        self.sma2 = self.I(SMA, self.data.Close, 200, overlay=True)
```
the initial 200 bars with undefined "SMA200c" will get skipped automatically. Why not do it like that?

**Reply by mikeFyngers** [reply]
Yeah, that's what I took originally from the tutorial. 

The reason is that I want to prescribe the backtest start date and (less important) the "buy and hold return stat" to be meaningful. I guess for the former I'll start the backtest 200 days earlier (if my indicator needs 200 days warm-up) and ignore the latter. This is also a fair solution, and better than the precomputation marked as preferred answer by tubular2000.

**Comment by zillionare** [comment]
the warm up period should be excluded from `Buy & Hold Return` calculation also.

**Comment by AGG2017** [comment]
We must have optional parameters date_begin and date_end to run the backtesting. Now I'm trying to implement this because I badly need it as well as the trading orders over the options prices provided with a custom function from a database. I see that statistics use first and last value for the closing price from the prices dataframe to calculate the duration of the trades and the corresponding Buy & Hold Return, etc. This way if you have let say using Wave C for the John Carter's Squeeze system then the backtesting will skip first ~400 bars for the trading but they will be included for the Buy & Hold Return which is wrong. We must define the starting date for everything, otherwise we will never create a strategy close enough to the wrongly calculated Buy & Hold Return with 2 years (in my case) of advance calculation.

**Reply by kernc** [reply]
This is already tracked as issue https://github.com/kernc/backtesting.py/issues/327. The acceptable solution would be to automatically clip Buy&Hold in stats to the active period.

**Comment by AGG2017** [comment]
I didn't see this issue but I mention my observation just to accent how important is to have clear start and end for everything to avoid such mysteries how good is just to buy and hold.
I did some initial implementation and the easiest way for me was to add begin and end index parameters for the stat function. By default they are 0 and -1 to keep the function compatible. The run function accept date_begin and date_end as dates, skip the strategy processing outside of this date range and convert these dates as begin/end index that are transferred to stat for proper calculations. Everything looks good, clean and easy so far but I have to do more testing to be sure I don't miss anything.
